src/library_vr.js

```diff
@@ -1,237 +1,373 @@
-        return;
+        WebVR.displays = [];
+        return 0;
       }
 
-      navigator.getVRDevices().then(
-        function(devs) {
-          WebVR.ready = true;
-          WebVR.devices = devs;
-        },
-        function() {
-          console.log("Emscripten WebVR getVRDevices() hit error callback");
-          WebVR.ready = true;
-          WebVR.devices = [];
-        }
-      );
+      WebVR.version = [1, 1];
+
+      navigator.getVRDisplays().then(function(displays) {
+        WebVR.ready = true;
+        WebVR.displays = displays;
+      });
+
+      return 1;
     },
 
-    getDeviceByID: function(deviceId) {
-      if (deviceId < 1 || deviceId > WebVR.devices.length) {
-        console.log("library_vr getDeviceByID invalid device id at: " + stackTrace());
+    dereferenceDisplayHandle: function(displayHandle) {
+      /* Display handles start as 1 as 0 will be interpreted as false or null-handle
+       * on errors */
+      if (displayHandle < 1 || displayHandle > WebVR.displays.length) {
+        console.log("library_vr dereferenceDisplayHandle invalid display handle at: " + stackTrace());
         return null;
       }
 
-      return WebVR.devices[deviceId-1];
+      return WebVR.displays[displayHandle-1];
     }
   },
 
   emscripten_vr_init: function() {
-    WebVR.init();
+    return WebVR.init();
+  },
+
+  emscripten_vr_version_major: function() {
+    return WebVR.version[0];
+  },
+
+  emscripten_vr_version_minor: function() {
+    return WebVR.version[1];
   },
 
   emscripten_vr_ready: function() {
     return WebVR.ready ? 1 : 0;
   },
 
-  emscripten_vr_count_devices: function() {
-    return WebVR.devices.length;
+  emscripten_vr_count_displays: function() {
+    return WebVR.displays.length;
   },
 
-  emscripten_vr_get_device_id: function(deviceIndex) {
-    if (deviceIndex < 0 || deviceIndex >= WebVR.devices.length) {
+  emscripten_vr_get_display_handle: function(displayIndex) {
+    if (displayIndex < 0 || displayIndex >= WebVR.displays.length) {
       return -1;
     }
 
-    // we're doing to treat the device ID the same as the index + 1
-    return deviceIndex + 1;
+    /* As displayHandle == 0 will be interpreted as NULL handle for errors,
+     * the handle is index + 1. */
+    return displayIndex + 1;
   },
 
-  emscripten_vr_get_device_hwid: function(deviceId) {
-    var dev = WebVR.getDeviceByID(deviceId);
-    if (!dev) return -1;
+  emscripten_vr_get_display_name: function(displayHandle) {
+    var display = WebVR.dereferenceDisplayHandle(displayHandle);
 
-    if (!WebVR.deviceHardwareIds[dev.hardwareUnitId]) {
-      WebVR.deviceHardwareIds[dev.hardwareUnitId] = WebVR.nextHardwareDeviceId++;
-    }
-    return WebVR.deviceHardwareIds[dev.hardwareUnitId];
+    var buffer, displayName;
+    displayName = display ? display.displayName : "";
+    var len = lengthBytesUTF8(displayName);
+    buffer = _malloc(len + 1);
+    stringToUTF8(displayName, buffer, len + 1);
+
+    return buffer;
```

Comment:
It looks like we have a memory leak here, the above code `malloc`s, and responsibility to `free` is assigned to the caller, but the documentation does not state that. I think rather than assigning burden to `free` to caller, it will be better to statically allocate on first call, and hold on to the allocation, returning the same allocated pointer each time, and WebVR subsystem deinit will then eventually `free` the allocation.

Or alternatively, user passes in the memory block where to write to, but that is probably unneedlessly complicated, so the current signature might be preferred, since these are expected to be static.

Reply:
I like the static string, as this is how `glGetString()` does things. I also added `emscripten_vr_deinit()` along with it.